### DISCUSSION TITLE
Singular matrix with pressure control

### DISCUSSION THREAD

**Discussion by dad616610** [e2nIEE/pandapipes#626] [Q&A]
Hello!

I'm getting a singular matrix when using pressure control with the following initial conditions: mass flow at inlet (`source`) and pressure at outlet (`ext_grid`). I tested this behavior with replacing the pressure control with pipe - no singular matrix, got solution. I also changed the initial conditions to pressure at inlet (`ext_grid`) and mass flow at outlet (`sink`) and ran with both pressure control and pipe - got solutions in both cases. 

Here is a table with the results: 
| | flow, press | press, flow |
|--------|--------|--------|
| press_control | False | True |
| pipe | True | True |

Table legend:
* "flow, press" =  inlet flow, outlet pressure; "press, flow" - inlet pressure, outlet flow
*  press_control = pressure control was used; pipe = pipe was used instead of pressure control
* `False` = got singular matrix; `True` - no singular matrix. got solution

<details><summary>Here is the code, simulating all the configurations above</summary>
<p>

```py
import pandapipes as pp


def main(init_conds, use_press_control):
    net = pp.create_empty_network("network", fluid="lgas")

    pp.create_junctions(net, nr_junctions=6, pn_bar=0.8, tfluid_k=300)

    pp.create_pipes_from_parameters(
        net,
        from_junctions=[0, 1, 3, 4],
        to_junctions=[1, 2, 4, 5],
        length_km=0.1,
        diameter_m=0.2,
    )

    if use_press_control:
        pp.create_pressure_control(
            net,
            from_junction=2,
            to_junction=3,
            controlled_junction=3,
            controlled_p_bar=0.5,
        )
    else:
        pp.create_pipe_from_parameters(
            net,
            from_junction=2,
            to_junction=3,
            length_km=0.1,
            diameter_m=0.2,
        )

    if init_conds == "press_flow":
        pp.create_ext_grid(net, junction=0, p_bar=0.5)
        pp.create_sink(net, junction=5, mdot_kg_per_s=0.06)
    elif init_conds == "flow_press":
        pp.create_source(net, junction=0, mdot_kg_per_s=0.06)
        pp.create_ext_grid(net, junction=5, p_bar=0.5)

    pp.pipeflow(net, friction_model="colebrook")


if __name__ == "__main__":
    from itertools import product

    for init_conds, use_press_control in product(
        ("press_flow", "flow_press"),
        (True, False),
    ):
        try:
            main(init_conds, use_press_control)
            res = "Success!"
        except pp.PipeflowNotConverged:
            res = "Fail!"
        finally:
            print(f"{init_conds = }, {use_press_control = }\n{res}")

```

</p>
</details> 

Could I be doing something wrong?

**Comment by jkisse** [comment]
Hi @dad616610, 
sorry for the late reply. Please be aware the pressure control kind of splits your network into two sections, upstream and downstream. In both sections, at least one junction needs a fixed pressure. Downstream, this is achieved by the pressure controller's `controlled_junction`. Upstream, you should add an ext. grid (with a pressure higher than the pressure controller's `controlled_p_bar`, so that the pressure gradient allows for a flow into the pressure controller). 

Note that the pressure of the "uncontrolled" junctions downstream of the pressure controller is determined by the hydraulics of the network, i.e. the pressure loss due to friction etc.. If there is an additional ext_grid there, trying to impose a certain pressure on such a junction, you'll likely end up with an overdetermined network (--> will typically raise "matrix is exactly singular").

I hope this helps :)

**Reply by lsc-ysu** [reply]
Hi, thank you for the explanation. That clarifies why the singular matrix occurs.

  In my case, the medium-pressure network can also be supplied by a local gas source. I initially modeled this local source as `ext_grid(type="p")`, because the station/source pressure is known. I now
  understand that, if this `ext_grid` is located downstream of a `press_control`, it imposes an additional fixed-pressure boundary and can overdetermine the downstream section.

  For this kind of use case, what would be the recommended modeling approach in pandapipes?

  1. Model the local medium-pressure source as a `source` with prescribed `mdot_kg_per_s`, if the injection flow is known.
  2. Model the local source as an `ext_grid(type="p")` at a separate source junction, connected to the medium-pressure network through a pipe/valve/equivalent resistance, if the source pressure is known but
  the injection flow should be determined by hydraulics.
  3. Implement external control logic that switches the `press_control` in or out of service depending on the downstream pressure, similar to a real regulator closing when the downstream pressure is already
  above the set point.

  Also, would it make sense for pandapipes to detect this situation earlier and raise a more explicit warning/error instead of only showing `MatrixRankWarning: Matrix is exactly singular`? For example, a
  message indicating that the downstream section of a pressure controller contains another fixed-pressure boundary would make the issue much easier to diagnose.

  Thanks again for your help.
